ampd/src/cosmos.rs

```diff
@@ -0,0 +1,100 @@
+use async_trait::async_trait;
+use cosmrs::proto::cosmos::auth::v1beta1::query_client::QueryClient as AuthQueryClient;
+use cosmrs::proto::cosmos::auth::v1beta1::{QueryAccountRequest, QueryAccountResponse};
+use cosmrs::proto::cosmos::bank::v1beta1::query_client::QueryClient as BankQueryClient;
+use cosmrs::proto::cosmos::bank::v1beta1::{QueryBalanceRequest, QueryBalanceResponse};
+use cosmrs::proto::cosmos::tx::v1beta1::service_client::ServiceClient;
+use cosmrs::proto::cosmos::tx::v1beta1::{
+    BroadcastTxRequest, BroadcastTxResponse, GetTxRequest, GetTxResponse, SimulateRequest,
+    SimulateResponse,
+};
+use mockall::automock;
+use report::ErrorExt;
+use thiserror::Error;
+use tonic::transport::Channel;
+use tonic::Response;
+
+type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to connect to the grpc endpoint")]
+    GrpcConnection(#[from] tonic::transport::Error),
+    #[error("failed to make the grpc request")]
+    GrpcRequest(#[from] tonic::Status),
+}
+
+#[automock]
+#[async_trait]
+pub trait CosmosClient {
+    async fn broadcast_tx(&mut self, req: BroadcastTxRequest) -> Result<BroadcastTxResponse>;
+    async fn simulate(&mut self, req: SimulateRequest) -> Result<SimulateResponse>;
+    async fn tx(&mut self, req: GetTxRequest) -> Result<GetTxResponse>;
+
+    async fn account(&mut self, address: QueryAccountRequest) -> Result<QueryAccountResponse>;
+
+    async fn balance(&mut self, request: QueryBalanceRequest) -> Result<QueryBalanceResponse>;
+}
+
+#[derive(Clone)]
+pub struct CosmosGRpcClient {
+    auth: AuthQueryClient<Channel>,
+    bank: BankQueryClient<Channel>,
+    service: ServiceClient<Channel>,
+}
+
+impl CosmosGRpcClient {
+    pub async fn new(url: &str) -> Result<Self> {
+        let endpoint: tonic::transport::Endpoint = url.parse().map_err(ErrorExt::into_report)?;
```

Comment:
conceptual question: If we keep the client a slim facade that essentially just converts errors I'm ok keeping it untested in unit tests. Otherwise, we should think about mocking the endpoint instead of the client itself

Reply:
This client is gonna be kept as simple as it is now. You'll see what I did in my next PR and how things are tested.